ee/clickhouse/sql/trends/lifecycle.py

```diff
@@ -1,184 +1,99 @@
 
-    FROM events
-        JOIN ({GET_TEAM_PERSON_DISTINCT_IDS}) pdi 
-            ON events.distinct_id = pdi.distinct_id
+        However, we want to have something of the format:
 
-    WHERE 
-        team_id = %(team_id)s 
-        AND {event_query} {filters}
-        AND toDateTime(events.timestamp) <= toDateTime(%(date_to)s) 
-        AND {trunc_func}(events.timestamp) >= toDateTime(%(date_from)s)
-        
-    GROUP BY person_id
-), periods AS (
-    SELECT 
-        {trunc_func}(toDateTime(%(date_to)s) - number * %(seconds_in_interval)s) AS start_of_period 
-        
-    FROM numbers(%(num_intervals)s)
-)
+        person_id  | period_of_activity          |  status_of_activity
+        person_id  | period_just_after_activity  |  dormant_status_of_period_after_activity
 
-SELECT 
-    activity_pairs.person_id AS person_id,
-    activity_pairs.initial_period AS initial_period,
-    activity_pairs.next_period AS next_period, 
-    if(
-        initial_period = toDateTime('0000-00-00 00:00:00'), 
-        'dormant', 
+        such that we can simply aggregate over person_id, period.
+    */
+    arrayJoin(
+        arrayZip(
+            [period, period + INTERVAL 1 {interval_expr}],
+            [initial_status, if(next_is_active, '', 'dormant')]
+        )
+    ) AS period_status_pairs,
+    period_status_pairs.1 as start_of_period,
+    period_status_pairs.2 as status
+FROM (
+    SELECT
+        person_id,
+        period,
+        created_at,
         if(
-            next_period = initial_period + INTERVAL {interval}, 
-            'returning', 
+            dateTrunc(%(interval)s, created_at) = period,
+            'new',
             if(
-                next_period > earliest + INTERVAL {interval}, 
-                'resurrecting', 
-                'new'
+                previous_activity + INTERVAL 1 {interval_expr} = period,
+                'returning',
+                'resurrecting'
             )
-        )
-    ) as status
-
-FROM (
-    /*
-         Get person period activity paired with the next adjacent period activity
-    */
-    SELECT 
-        person_id, 
-        initial_period, 
-        min(next_period) as next_period 
-
+        ) AS initial_status,
+        period + INTERVAL 1 {interval_expr} = following_activity AS next_is_active,
+        previous_activity,
+        following_activity
     FROM (
-        SELECT 
-            person_id, 
-            base.start_of_period as initial_period, 
-            subsequent.start_of_period as next_period
-
-        FROM person_activity_including_previous_period base
-            JOIN person_activity_including_previous_period subsequent 
-                ON base.person_id = subsequent.person_id
-
-        WHERE subsequent.start_of_period > base.start_of_period
+        SELECT
+            person_id,
+            any(period) OVER (PARTITION BY person_id ORDER BY period ROWS BETWEEN 1 PRECEDING AND 1 PRECEDING) as previous_activity,
```

Comment:
Love it, does this mean we only have one JOIN on pdis/persons? That should be a massive save on memory I guess and speed.

Reply:
Yup. This can backfire though - in a sharded setting, I'm assuming there will need to be some back-and-forth communication between nodes.